[12.x] Fix accessing `Connection` property in `Grammar` classes

hafezdivandari opened the pull request:
The `Illuminate\Database\Grammar` class has required a `Connection $connection` property since Laravel 10.x (via #46558), but it was passed via `setConnection()` method to avoid signature changes in a patch release. Since then, many methods in the `Grammar` classes have needed access to the `Connection` property (mainly to retrieve config properties and compare driver versions). This PR does:

* Adding `Connection $connection` to the `Grammar` constructor, instead of calling `setConnection()` when instantiating.
* Fixing table prefix issues by using `Connection::getTablePrefix()` instead of having the `$prefix` property in the `Grammar` and `Blueprint` classes.

**P.S. 1**: This PR doesn’t require any upgrade guide for end-users.

**P.S. 2**: The `$connection` property was added to the `Blueprint` constructor in Laravel 12.x via #51821, so now is a good time to fix it in the `Grammar` class as well.

## Summary

* Adds missing `prefix_indexes` config property to `sqlite` connection to be consistent with other DB driver connections.
* Adds `Connection $connection` property to `Illuminate/Database/Grammar::__construct()` method.
* Removes `Illuminate\Database\Connection::withTablePrefix()` method.
* Removes `Illuminate\Database\Grammar::setConnection()` method.
* Removes `Illuminate\Database\Schema\Builder::setConnection()` method.
* Removes `$prefix` property from `Illuminate\Database\Schema\Blueprint::__construct()` method.
* Deprecates `Illuminate\Database\Grammar::getTablePrefix()` method (Use `Illuminate\Database\Connection::getTablePrefix()` instead).
* Deprecates `Illuminate\Database\Grammar::setTablePrefix()` method (Use `Illuminate\Database\Connection::setTablePrefix()` instead).
* Deprecates `Illuminate\Database\Schema\Blueprint::getPrefix()` method (Use `Illuminate\Database\Connection::getTablePrefix()` instead).
* Fix a bug on wrapping aliased tables with prefix.
* Fix a bug where table prefix was totally ignored on `Blueprint` when `prefix_indexes` is set to `false`.
* Fix a bug on defining starting value for a column of a table with non-default schema on PostgreSQL.
* Moves overridden `createDatabase()` and `dropDatabaseIfExists()` methods to `Illuminate\Database\Schema\Builder` parent class (The same method was overridden on all inherited classes.)

hafezdivandari committed:
fix connection references

hafezdivandari committed:
formatting

Jubeki commented on src/Illuminate/Database/Grammar.php:
```diff
@@ -284,35 +287,26 @@ public function getDateFormat()
     /**
      * Set the grammar's table prefix.
      *
+     * @deprected DB::setTablePrefix()
```
```suggestion
     * @deprecated DB::setTablePrefix()
```

Jubeki commented on src/Illuminate/Database/Grammar.php:
```diff
@@ -284,35 +287,26 @@ public function getDateFormat()
     /**
      * Get the grammar's table prefix.
      *
+     * @deprected Use DB::getTablePrefix()
```
```suggestion
     * @deprecated Use DB::getTablePrefix()
```

Jubeki commented on src/Illuminate/Database/Schema/Blueprint.php:
```diff
@@ -1824,11 +1819,13 @@ public function getTable()
     /**
      * Get the table prefix.
      *
+     * @deprected Use DB::getTablePrefix()
```
```suggestion
     * @deprecated Use DB::getTablePrefix()
```

hafezdivandari committed:
formatting

hafezdivandari committed:
formatting

tpetry commented:
I've implemented passing `Connection` as a parameter and not a constructor value to simplify work for package maintainers. There are many packages that extend the different Eloquent drivers to add some stuff which is not available in Laravel.

Moving the `Connection` to the constructor makes writing packages that target multiple Laravel versions impossible. Thats why I did it that way - and would prefer it being kept like that 😉 My PG driver for example is compatible back to Laravel 6 (!) because of that design.

hafezdivandari commented:
@tpetry Thank you for reviewing and dislike😅. But this PR isn't just a styling fix, accessing the `connection` on `Grammar` classes needs to be addressed sooner or later IMHO. However:

1. As I mentioned in the PR description, passing `$connection` to the `Blueprint` constructor has already been merged in #51821. So, the signature changes are already unavoidable?

2. The current implemented way of passing `$connection` to `Grammar` classes is problematic, leading to hidden bugs and decoupling issues. For example, this PR addresses issues like accessing/manipulating the table prefix. Here's a sample from your package (amazing work by the way): [laravel-postgresql-enhanced/src/Schema/Grammars/GrammarTypes.php#L20-L23](https://github.com/tpetry/laravel-postgresql-enhanced/blob/e952798248f643920c650e813bc10e0258dde304/src/Schema/Grammars/GrammarTypes.php#L20-L23)

3. This change helps avoid BC breaks in the future. For example, the intention of #51821 was a small change, but needing access to `Connection` made it large and hard to review, it could have been done in a patch release!

4. Code style fixers complain about accessing `$this->connection` in `Grammar` classes, as the property may be `null`.

tpetry commented:
Not a big fan, but the package authors will find a way to make it work with the bc break here.
